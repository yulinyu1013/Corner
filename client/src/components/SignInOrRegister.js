import React, { useState } from 'react';
import '../styles/signIn.css';
import '../styles/registration.css';
import Cookies from 'js-cookie';
import { register, login } from '../fetchers/signInAndRegister';
import {
  usernameValidation,
  emailValidation,
  passwordValidation,
  passwordConfirmValidation,
} from '../helpers/userInfoValidationHelper';

function SignInOrRegister({ setIsLoggedIn, setCurrentUser }) {
  const [signIn, setSignIn] = useState(true);
  const [signInUsername, setSignInUsername] = useState('');
  const [signInPassword, setSignInPassword] = useState('');
  const [signInValidation, setSignInValidation] = useState('');

  const [registerUsername, setRegisterUsername] = useState('');
  const [registerEmail, setRegisterEmail] = useState('');
  const [registerPassword, setRegisterPassword] = useState('');
  const [registerPasswordConfirm, setRegisterPasswordConfirm] = useState('');

  const [registerUsernameValidation, setRegisterUsernameValidation] = useState('');
  const [registerEmailValidation, setRegisterEmailValidation] = useState('');
  const [registerPasswordValidation, setRegisterPasswordValidation] = useState('');
  const [registerPasswordConfirmValidation, setRegisterPasswordConfirmValidation] = useState('');

  // lockout

  // redirect handlers
  const redirectToRegister = (e) => {
    e.preventDefault();
    setSignIn(false);
  };

  const redirectToSignIn = (e) => {
    e.preventDefault();
    setSignIn(true);
  };

  // sign in handlers
  const signInUsernameChangeHandler = (e) => {
    setSignInUsername(e.target.value);
  };

  const signInPasswordChangeHandler = (e) => {
    setSignInPassword(e.target.value);
  };

  // const history = useHistory();

  const signInHandler = async (e) => {
    e.preventDefault();
    // check if account is locked
    if (Cookies.get(signInUsername)) {
      setSignInValidation('Account locked due to multiple failed login attempts. Please try again later.');
      return;
    }
    const signInInfo = { name: signInUsername, password: signInPassword };
    login(signInInfo).then((res) => {
      const currentUser = res.data;
      console.log(currentUser);
      const tempStorage = {
        id: currentUser.id,
        name: currentUser.name,
        bio: currentUser.bio,
        avatar: currentUser.avatar,
      };
      sessionStorage.setItem('userInfo', JSON.stringify(tempStorage));
      // history.push(`/${currentUser.id}`);
      console.log(window.location.href);
      setCurrentUser(currentUser);
      setIsLoggedIn();
    }).catch((err) => {
      console.log(err);
      const currTime = Date.now();
      if (!sessionStorage.getItem(`${signInUsername}-loginAttempts`)) {
        sessionStorage.setItem(`${signInUsername}-loginAttempts`, JSON.stringify({ num: 1, first: currTime, last: currTime }));
        setSignInValidation('Invalid Username or Password.');
        return;
      }

      let attemptsInfo = JSON.parse(sessionStorage.getItem(`${signInUsername}-loginAttempts`));
      // if over 5 min since last attempt - treated as a new attempt
      if (currTime - attemptsInfo.first > 5 * 60 * 1000) {
        attemptsInfo = { num: 1, first: currTime, last: currTime };
      } else {
        // else - update attempts
        attemptsInfo = { num: attemptsInfo.num + 1, first: attemptsInfo.first, last: currTime };
      }
      sessionStorage.setItem(`${signInUsername}-loginAttempts`, JSON.stringify(attemptsInfo));

      if (attemptsInfo.num < 3) {
        setSignInValidation('Invalid Username or Password.');
      } else if (attemptsInfo.num === 3) {
        Cookies.set(signInUsername, 'failed', { expires: Date.now() + (30 * 60 * 1000) });
        setSignInValidation('3 failed login attempts in 5 min. Please try 30 min later.');
        sessionStorage.removeItem(`${signInUsername}-loginAttempts`);
      }
    });
  };

  // registration handlers
  const registerUsernameHandler = (e) => {
    setRegisterUsername(e.target.value);
    setRegisterUsernameValidation(usernameValidation(e.target.value));
  };

  const registerEmailHandler = (e) => {
    setRegisterEmail(e.target.value);
    setRegisterEmailValidation(emailValidation(e.target.value));
  };

  const registerPasswordHandler = (e) => {
    setRegisterPassword(e.target.value);
    setRegisterPasswordValidation(passwordValidation(e.target.value));
  };

  const registerPasswordConfirmHandler = (e) => {
    const { value } = e.target;
    setRegisterPasswordConfirm(value);
    setRegisterPasswordConfirmValidation(passwordConfirmValidation(registerPassword, value));
  };

  const registerHandler = async () => {
    const newUser = { name: registerUsername, email: registerEmail, password: registerPassword };
    if (registerUsernameValidation === ''
      && registerEmailValidation === ''
      && registerPasswordValidation === ''
      && registerPasswordConfirmValidation === '') {
      register(newUser).then((res) => {
        const currentUser = res.data;
        console.log(currentUser);
        const tempStorage = {
          id: currentUser.id,
          name: currentUser.name,
          bio: currentUser.bio,
          avatar: currentUser.avatar,
        };
        sessionStorage.setItem('userInfo', JSON.stringify(tempStorage));
        // history.push(`/${currentUser.id}`);
        setCurrentUser(currentUser);
        setIsLoggedIn();
      }).catch((err) => {
        if (err.message === 'Request failed with status code 409') {
          setRegisterUsernameValidation('Username has already been used.');
        }
      });
    }
  };

  return (
    <div className="signInOrRegisterContainer" data-testid="signIn">
      {signIn
        ? (
          <div className="signIn_container">
            <div className="signIn_welcome">
              Welcome
              <br />
              Back to
            </div>
            <div className="signIn_corner"> Corner. </div>
            <div className="signIn_welcomeMsg">
              Get relax &amp; get enlightened
              <br />
              with bookðŸ“š, movieðŸŽ¬, musicðŸŽµ
              <br />
              and people.
            </div>
            <div className="signInMsg">Sign in your account below.</div>
            <div className="signIn_input">
              <input className="signIn_username" type="text" name="username" value={signInUsername} placeholder="Username" onChange={signInUsernameChangeHandler} />
              <input className="signIn_password" type="password" name="password" value={signInPassword} placeholder="Password" onChange={signInPasswordChangeHandler} />
            </div>
            <div className="signIn_validation" data-testid="test">{signInValidation}</div>
            <button className="signIn_button" type="submit" onClick={signInHandler} disabled={!signInUsername || !signInPassword}> Sign In </button>
            <div className="register_redirect_div">
              <div className="register_redirect">
                Don&apos;t an account? &nbsp;
                <a href="/" className="register_redirect_link" data-testid="to-register" onClick={redirectToRegister}>Register</a>
              </div>
            </div>
          </div>
        )
        : (
          <div className="registeration-container" data-testid="registration">
            <div className="register-welcome"> Welcome to </div>
            <div className="register-corner"> Corner. </div>
            <div className="register-welcomeMsg">
              A cozy space to
              <br />
              relax &amp; get enlightened
              <br />
              with bookðŸ“š, movieðŸŽ¬, musicðŸŽµ
              <br />
              and people.
            </div>
            <div className="registerMsg">Create your account below.</div>
            <div className="register-input">
              <div className="registration-validation">{registerUsernameValidation}</div>
              <input className="register-username" type="text" name="username" placeholder="Username" value={registerUsername} onChange={registerUsernameHandler} />
              <div className="registration-validation">{registerEmailValidation}</div>
              <input className="register-email" type="email" name="email" placeholder="Email" value={registerEmail} onChange={registerEmailHandler} />
              <div className="registration-validation">{registerPasswordValidation}</div>
              <input className="register-password" type="password" name="password" placeholder="Password" value={registerPassword} onChange={registerPasswordHandler} />
              <div className="registration-validation">{registerPasswordConfirmValidation}</div>
              <input className="register-password_comfirm" type="password" name="password_comfirm" placeholder="Comfirm Password" value={registerPasswordConfirm} onChange={registerPasswordConfirmHandler} />
            </div>
            <button className="register-button" type="submit" onClick={registerHandler} disabled={!registerUsername || !registerEmail || !registerPassword || !registerPasswordConfirm}> Register </button>
            <div className="sign_in_redirect">
              Already have an account?&nbsp;
              <a href="/" className="sign_in_redirect_button" data-testid="to-signIn" onClick={redirectToSignIn}>Sign In</a>
            </div>
          </div>
        )}
    </div>
  );
}

export default SignInOrRegister;
