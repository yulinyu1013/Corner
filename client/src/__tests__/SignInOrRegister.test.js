import React from 'react';
import SignInOrRegister from '../components/SignInOrRegister';
import renderer from 'react-test-renderer';
import {act, render, fireEvent, cleanup, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
const fetcherHelpers = require('../fetchers/signInAndRegister');
jest.mock('../fetchers/signInAndRegister');


describe('UI', () => {
  it('sign in page', () => {
    render(<SignInOrRegister />);
    const linkElement = screen.getByTestId('signIn');
    expect(linkElement).toBeInTheDocument();
  });

  it('redirect to register page', () => {
    render(<SignInOrRegister />);
    const RegisterButton = screen.getByTestId('to-register');
    fireEvent.click(RegisterButton);
    const register = screen.getByTestId('registration');
    expect(register).toBeInTheDocument();
  });

  it('back to the sign in page', () => {
    render(<SignInOrRegister />);
    const RegisterButton = screen.getByTestId('to-register');
    fireEvent.click(RegisterButton);
    const signInButton = screen.getByTestId('to-signIn');
    fireEvent.click(signInButton);
    const signInAgain = screen.getByTestId('signIn');
    expect(signInAgain).toBeInTheDocument();
  });


})

describe('SignIn', () => {

  afterEach(cleanup);

  it('enter username', () => {
    render(<SignInOrRegister />);
    userEvent.type(screen.getByPlaceholderText('Username'), 'testUsername');
    expect(screen.getByPlaceholderText('Username')).toHaveValue('testUsername')
  })

  it('enter password', () => {
    render(<SignInOrRegister />);
    userEvent.type(screen.getByPlaceholderText('Password'), 'testPassword');
    expect(screen.getByPlaceholderText('Password')).toHaveValue('testPassword');
  })

  it('sign in successfully', async () => {
    const mockUser = {
      name: 'testUsername', 
      bio: 'test bio', 
      email: 'test@gmail.com', 
      password: 'testPassword', avatar: null,
    }
    jest.spyOn(fetcherHelpers, 'login').mockResolvedValue({ data: mockUser });
   
    const setCurrentUser = jest.fn();
    const setIsLoggedIn = jest.fn();
    const setSignInValidation = jest.fn();
    render(<SignInOrRegister setCurrentUser={setCurrentUser} setIsLoggedIn={setIsLoggedIn} setSignInValidation={setSignInValidation}/>);
    userEvent.type(screen.getByPlaceholderText('Username'), 'testUsername');
    userEvent.type(screen.getByPlaceholderText('Password'), 'testPassword');
    act(() => {
      fireEvent.click(screen.getByText('Sign In'));
    })
    await waitFor(() => expect(setCurrentUser).toHaveBeenCalled());
    await waitFor(() => expect(setIsLoggedIn).toHaveBeenCalled());
    await waitFor(() => expect(setSignInValidation).not.toHaveBeenCalled());

  })

  it('sign in unsuccessfully', async () => {
  
    jest.spyOn(fetcherHelpers, 'login').mockRejectedValue('log in failed');
    
    const setCurrentUser = jest.fn();
    const setIsLoggedIn = jest.fn();
    const setSignInValidation = jest.fn();
    render(<SignInOrRegister setCurrentUser={setCurrentUser} setIsLoggedIn={setIsLoggedIn} setSignInValidation={setSignInValidation}/>);
    userEvent.type(screen.getByPlaceholderText('Username'), 'wrongUsername');
    userEvent.type(screen.getByPlaceholderText('Password'), 'wrongPassword');
    act(() => {
      fireEvent.click(screen.getByText('Sign In'));
    })
    
    await waitFor(() => expect(screen.getByText('Invalid Username or Password.')).toBeInTheDocument); 
  })

})

describe('Register', () => {
  afterEach(cleanup);

  it('enter username', () => {
    render(<SignInOrRegister />);
    fireEvent.click(screen.getByTestId('to-register'));
    userEvent.type(screen.getByPlaceholderText('Username'), 'testUsername');
    expect(screen.getByPlaceholderText('Username')).toHaveValue('testUsername')
  })

  it('enter email', () => {
    render(<SignInOrRegister />);
    fireEvent.click(screen.getByTestId('to-register'));
    userEvent.type(screen.getByPlaceholderText('Email'), 'test@gmail.com');
    expect(screen.getByPlaceholderText('Email')).toHaveValue('test@gmail.com');
  })

  it('enter password', () => {
    render(<SignInOrRegister />);
    fireEvent.click(screen.getByTestId('to-register'));
    userEvent.type(screen.getByPlaceholderText('Password'), 'testPassword');
    expect(screen.getByPlaceholderText('Password')).toHaveValue('testPassword');
  })

  it('enter password confirm', () => {
    render(<SignInOrRegister />);
    fireEvent.click(screen.getByTestId('to-register'));
    userEvent.type(screen.getByPlaceholderText('Comfirm Password'), 'testPassword');
    expect(screen.getByPlaceholderText('Comfirm Password')).toHaveValue('testPassword');
  })

  it('register successfully', async () => {
    const mockNewUser = {
      name: 'testUsername',
      email: 'test@gmail.com',
      password: 'testPassword'
    }
    jest.spyOn(fetcherHelpers, 'register').mockResolvedValue({ data: mockNewUser });
    
    act(() => {
      render(<SignInOrRegister setSignIn={jest.fn()}/>);
    })
    fireEvent.click(screen.getByTestId('to-register'));
    userEvent.type(screen.getByPlaceholderText('Username'), 'testUsername');
    userEvent.type(screen.getByPlaceholderText('Email'), 'test@gmail.com');
    userEvent.type(screen.getByPlaceholderText('Password'), 'testPassword');
    userEvent.type(screen.getByPlaceholderText('Comfirm Password'), 'testPassword');
    act(() => {
      fireEvent.click(screen.getByText('Register'));
    })
    const linkElement = screen.getByTestId('signIn');
    expect(linkElement).toBeInTheDocument();

  })
})

it('renders SignInOrRegister correctly', () => {
  const tree = renderer.create(<SignInOrRegister />).toJSON();
  expect(tree).toMatchSnapshot();
});
