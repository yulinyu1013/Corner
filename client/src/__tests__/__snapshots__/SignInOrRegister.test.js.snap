// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SignInOrRegister correctly 1`] = `
<div
  className="signInOrRegisterContainer"
  data-testid="signIn"
>
  <div
    className="signIn_container"
  >
    <div
      className="signIn_welcome"
    >
      Welcome
      <br />
      Back to
    </div>
    <div
      className="signIn_corner"
    >
       Corner. 
    </div>
    <div
      className="signIn_welcomeMsg"
    >
      Get relax & get enlightened
      <br />
      with bookðŸ“š, movieðŸŽ¬, musicðŸŽµ
      <br />
      and people.
    </div>
    <div
      className="signInMsg"
    >
      Sign in your account below.
    </div>
    <div
      className="signIn_input"
    >
      <input
        className="signIn_username"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        type="text"
        value=""
      />
      <input
        className="signIn_password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <div
      className="signIn_validation"
      data-testid="test"
    >
      
    </div>
    <button
      className="signIn_button"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
       Sign In 
    </button>
    <div
      className="register_redirect_div"
    >
      <div
        className="register_redirect"
      >
        Don't an account? Â 
        <a
          className="register_redirect_link"
          data-testid="to-register"
          href="/"
          onClick={[Function]}
        >
          Register
        </a>
      </div>
    </div>
  </div>
</div>
`;
